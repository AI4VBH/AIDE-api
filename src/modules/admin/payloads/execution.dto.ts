export class ExecutionDTO {
  execution_id: number;
  payload_id: number;
  model_name: string;
  execution_status: string;
  execution_started: string;
  execution_finished: string;
  executions: ExecutionDTO[];

  public static from(dto: Partial<ExecutionDTO>) {
    const it = new ExecutionDTO();
    it.execution_id = dto.execution_id;
    it.payload_id = dto.payload_id;
    it.model_name = dto.model_name;
    it.execution_status = dto.execution_status;
    it.execution_started = dto.execution_started;
    it.execution_finished = dto.execution_finished;
    it.executions = dto.executions;
    return it;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Execution) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(execution: Execution = null) {
  //   const it = new Execution();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
