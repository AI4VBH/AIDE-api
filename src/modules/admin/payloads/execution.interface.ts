export interface Execution {
  execution_id: number;
  payload_id: number;
  model_name: string;
  execution_status: string;
  execution_started: string;
  execution_finished: string;
  executions: Execution[];
}
