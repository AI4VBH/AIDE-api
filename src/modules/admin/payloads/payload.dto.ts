/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { IMonaiPayload } from './payload.interface';

export class PayloadDTO {
  payload_id: string;
  patient_name: string;
  patient_id: string;
  payload_received: string;

  public static from(dto: Partial<PayloadDTO>) {
    const it = new PayloadDTO();
    it.payload_id = dto.payload_id;
    it.patient_name = dto.patient_name;
    it.patient_id = dto.patient_id;
    it.payload_received = dto.payload_received;
    return it;
  }

  public static fromMonaiPayload(entity: IMonaiPayload) {
    return PayloadDTO.from({
      payload_id: entity.payload_id,
      patient_name: entity.patient_details.patient_name,
      patient_id: entity.patient_details.patient_id,
      payload_received: entity.timestamp,
    });
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Payload) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(payload: Payload = null) {
  //   const it = new Payload();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
