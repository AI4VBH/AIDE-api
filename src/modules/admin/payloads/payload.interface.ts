/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { IPagedMonaiResponse } from 'shared/helper/paging/paging.interface';

export interface IGetPayloadsQueryParams {
  pageNumber: number;
  pageSize: number;
}

export interface IMonaiPayloadResponse
  extends IPagedMonaiResponse<IMonaiPayload> {
  succeeded: boolean;
  errors: string[];
  message: string;
}

export interface IMonaiPayload {
  id: string;
  payload_id: string;
  workflows: string[];
  workflow_instance_ids: string[];
  file_count: number;
  correlation_id: string;
  bucket: string;
  calling_aetitle: string;
  called_aetitle: string;
  timestamp: string;
  files: any[];
  patient_details: IMonaiPayloadPatient;
}

export interface IMonaiPayloadPatient {
  patient_id: string;
  patient_name: string;
  patient_sex: string;
  patient_dob: string;
  patient_age: string;
  patient_hospital_id: string;
}

export interface WorkflowInstanceDto {
  id: string;
  ae_title: string;
  workflow_name: string;
  workflow_id: string;
  start_time: string;
  payload_id: string;
  status: string;
  tasks: TaskExecutionDto[];
}

export interface TaskExecutionDto {
  execution_id: string;
  payload_id: string;
  workflow_instance_id: string;
  task_start_time: string;
  task_id: string;
  status: string;
  next_task: TaskExecutionDto[];
}
