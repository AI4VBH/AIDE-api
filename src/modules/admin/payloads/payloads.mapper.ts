import {
  MonaiWorkflowInstance,
  MonaiWorkflowTask,
} from '../workflowinstances/workflowinstances.interface';
import { TaskExecutionDto, WorkflowInstanceDto } from './payload.interface';

export function mapWorkflowInstancesToExecutions(
  instances: MonaiWorkflowInstance[],
) {
  if (!instances) {
    throw new Error('instances is required');
  }

  if (!Array.isArray(instances)) {
    throw new Error('instances is not an array');
  }

  return instances.map((instance) => {
    const topLevel = instance.tasks.filter(
      (t) => t.previous_task_id === '' || !t.previous_task_id,
    );

    const nested = instance.tasks.filter((t) => !!t.previous_task_id);

    const instanceDto: WorkflowInstanceDto = {
      id: instance.id,
      ae_title: instance.ae_title,
      workflow_id: instance.workflow_id,
      payload_id: instance.payload_id,
      start_time: instance.start_time,
      status: instance.status,
      tasks: topLevel.map((t) =>
        mapTaskToTaskExecution(instance.payload_id, t, nested),
      ),
    };

    return instanceDto;
  });
}

export function mapTaskToTaskExecution(
  payload_id: string,
  task: MonaiWorkflowTask,
  tasks: MonaiWorkflowTask[],
): TaskExecutionDto {
  if (!task || !tasks) {
    throw new Error('task and tasks are required');
  }

  const children = tasks.filter((t) => t.previous_task_id === task.task_id);

  return {
    payload_id,
    execution_id: task.execution_id,
    workflow_instance_id: task.workflow_instance_id,
    task_start_time: task.task_start_time,
    task_id: task.task_id,
    status: task.status,
    next_task: children.map((t) =>
      mapTaskToTaskExecution(payload_id, t, tasks),
    ),
  };
}
