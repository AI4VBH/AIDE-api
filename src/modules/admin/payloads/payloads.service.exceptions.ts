export class PayloadNotFoundException extends Error {}

export class PayloadBadRequestException extends Error {}
