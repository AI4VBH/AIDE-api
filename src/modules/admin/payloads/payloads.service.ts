import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { lastValueFrom } from 'rxjs';
import PagingDTO from 'src/common/helper/paging/paging.dto';
import { IPagedResponse } from 'src/common/helper/paging/paging.interface';
import { ExecutionDTO } from './execution.dto';
import { PayloadDTO } from './payload.dto';
import {
  IMonaiPayload,
  IMonaiPayloadResponse,
  IGetPayloadsQueryParams,
} from './payload.interface';

@Injectable()
export class PayloadsService {
  constructor(private readonly httpService: HttpService) {}

  async getPayloads(
    query: IGetPayloadsQueryParams,
  ): Promise<IPagedResponse<PayloadDTO>> {
    const source = lastValueFrom(
      this.httpService.get<IMonaiPayloadResponse>(
        `payload?pageNumber=${query.pageNumber}&pageSize=${query.pageSize}`,
      ),
    );

    return await source.then((response) => {
      if (!response.data.succeeded) {
        throw new Error(
          response.data.errors?.join('\n') ?? response.data.message,
        );
      }

      return PagingDTO.fromMonaiPagedResponse<IMonaiPayload, PayloadDTO>(
        response.data,
        PayloadDTO.fromMonaiPayload,
      );
    });
  }

  getPayloadExecutions(payload_id): ExecutionDTO[] {
    switch (payload_id) {
      case 1:
        return [
          {
            execution_id: 1,
            payload_id: 1,
            model_name: 'Dicom Received',
            execution_status: 'success',
            execution_started: '20220516T141114',
            execution_finished: '20220516T141200',
            executions: [
              {
                execution_id: 2,
                payload_id: 1,
                model_name: '3D Fetal Body MRI',
                execution_status: 'success',
                execution_started: '20220516T141200',
                execution_finished: '20220516T150000',
                executions: [
                  {
                    execution_id: 2,
                    payload_id: 5,
                    model_name: 'Output Sent to Destination',
                    execution_status: 'success',
                    execution_started: '20220516T150000',
                    execution_finished: '20220516T150234',
                    executions: [],
                  },
                ],
              },
              {
                execution_id: 3,
                payload_id: 1,
                model_name: '3D Fetal Brain MRI',
                execution_status: 'success',
                execution_started: '20220516T141200',
                execution_finished: '20220516T151440',
                executions: [
                  {
                    execution_id: 2,
                    payload_id: 5,
                    model_name: 'Output Sent to Destination',
                    execution_status: 'success',
                    execution_started: '20220516T151440',
                    execution_finished: '20220516T151540',
                    executions: [],
                  },
                ],
              },
              {
                execution_id: 4,
                payload_id: 1,
                model_name: '3D Fetal Heart MRI',
                execution_status: 'success',
                execution_started: '20220516T141200',
                execution_finished: '20220516T150000',
                executions: [
                  {
                    execution_id: 2,
                    payload_id: 5,
                    model_name: 'Output Sent to Destination',
                    execution_status: 'success',
                    execution_started: '20220516T150000',
                    execution_finished: '20220516T152614',
                    executions: [],
                  },
                ],
              },
            ],
          },
        ];
      case 2:
        return [
          {
            execution_id: 1,
            payload_id: 2,
            model_name: 'Dicom Received',
            execution_status: 'success',
            execution_started: '20220526T050215',
            execution_finished: '20220526T050312',
            executions: [
              {
                execution_id: 2,
                payload_id: 2,
                model_name: 'Stroke Pathway',
                execution_status: 'success',
                execution_started: '20220526T050312',
                execution_finished: '20220526T052232',
                executions: [
                  {
                    execution_id: 2,
                    payload_id: 5,
                    model_name: 'Output Sent to Destination',
                    execution_status: 'success',
                    execution_started: '20220526T052232',
                    execution_finished: '20220526T052312',
                    executions: [],
                  },
                ],
              },
            ],
          },
        ];
      case 3:
        return [
          {
            execution_id: 1,
            payload_id: 3,
            model_name: 'Dicom Received',
            execution_status: 'success',
            execution_started: '20220611T060316',
            execution_finished: '20220611T060446',
            executions: [
              {
                execution_id: 2,
                payload_id: 3,
                model_name: 'Brainminer CT',
                execution_status: 'success',
                execution_started: '20220611T060446',
                execution_finished: '20220611T063326',
                executions: [
                  {
                    execution_id: 2,
                    payload_id: 5,
                    model_name: 'Output Sent to Destination',
                    execution_status: 'success',
                    execution_started: '20220611T063326',
                    execution_finished: '20220611T063446',
                    executions: [],
                  },
                ],
              },
            ],
          },
        ];
      case 4:
        return [
          {
            execution_id: 1,
            payload_id: 4,
            model_name: 'Dicom Received',
            execution_status: 'success',
            execution_started: '20220620T070417',
            execution_finished: '20220620T070527',
            executions: [
              {
                execution_id: 2,
                payload_id: 4,
                model_name: 'MR Spectroscopy',
                execution_status: 'success',
                execution_started: '20220620T070527',
                execution_finished: '20220620T073712',
                executions: [
                  {
                    execution_id: 2,
                    payload_id: 5,
                    model_name: 'Output Sent to Destination',
                    execution_status: 'success',
                    execution_started: '20220620T073712',
                    execution_finished: '20220620T073812',
                    executions: [],
                  },
                ],
              },
            ],
          },
        ];
      case 5:
        return [
          {
            execution_id: 1,
            payload_id: 5,
            model_name: 'Dicom Received',
            execution_status: 'success',
            execution_started: '20220701T080518',
            execution_finished: '20220701T080618',
            executions: [
              {
                execution_id: 2,
                payload_id: 5,
                model_name: '3D Fetal Body MRI',
                execution_status: 'success',
                execution_started: '20220701T080618',
                execution_finished: '20220701T085228',
                executions: [
                  {
                    execution_id: 4,
                    payload_id: 5,
                    model_name: 'Output Sent to Destination',
                    execution_status: 'success',
                    execution_started: '20220701T085228',
                    execution_finished: '20220701T085318',
                    executions: [],
                  },
                ],
              },
              {
                execution_id: 3,
                payload_id: 5,
                model_name: '3D Fetal Heart MRI',
                execution_status: 'error',
                execution_started: '20220701T080618',
                execution_finished: '20220701T080818',
                executions: [],
              },
            ],
          },
        ];

      default:
        return [];
    }
  }
}
