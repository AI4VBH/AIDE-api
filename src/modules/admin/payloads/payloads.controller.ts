import {
  BadRequestException,
  Controller,
  Get,
  Inject,
  Param,
  Query,
} from '@nestjs/common';
import { IPagedResponse } from 'common/helper/paging/paging.interface';
import { ExecutionDTO } from './execution.dto';
import { PayloadDTO } from './payload.dto';
import { IGetPayloadsQueryParams } from './payload.interface';
import { PayloadsService } from './payloads.service';

@Controller('payloads')
export class PayloadsController {
  @Inject(PayloadsService)
  private readonly appService: PayloadsService;

  @Get()
  async getPayloads(
    @Query() query: IGetPayloadsQueryParams,
  ): Promise<IPagedResponse<PayloadDTO>> {
    if (!query || !query.pageNumber || !query.pageSize) {
      throw new BadRequestException(
        'pageNumber and pageSize are both required query parameters.',
      );
    }

    if (query.pageNumber < 1 || query.pageSize < 1) {
      throw new BadRequestException(
        'pageNumber and pageSize must both be a minimum of 1.',
      );
    }

    return await this.appService.getPayloads(query);
  }

  @Get(':payload_id/executions')
  getPayloadExecutions(@Param('payload_id') payload_id): ExecutionDTO[] {
    return this.appService.getPayloadExecutions(payload_id);
  }
}
