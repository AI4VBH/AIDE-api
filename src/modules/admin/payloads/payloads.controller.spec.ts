import { createMock, DeepMocked } from '@golevelup/ts-jest';
import { BadRequestException } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { IGetPayloadsQueryParams } from './payload.interface';
import { PayloadsController } from './payloads.controller';
import { PayloadsService } from './payloads.service';

describe('PayloadsController', () => {
  let service: DeepMocked<PayloadsService>;
  let controller: PayloadsController;

  beforeEach(async () => {
    service = createMock<PayloadsService>();

    const module: TestingModule = await Test.createTestingModule({
      controllers: [PayloadsController],
      providers: [
        {
          provide: PayloadsService,
          useValue: service,
        },
      ],
    }).compile();

    controller = module.get<PayloadsController>(PayloadsController);
  });

  it('should be defined', () => expect(controller).toBeDefined());

  describe('getPayloads', () => {
    it.each([
      { pageNumber: '0', pageSize: '0' },
      { pageNumber: '0', pageSize: '10' },
    ])(
      'throws bad request when either query string parameter is not a minimum of 1: %s',
      async (queryParams: IGetPayloadsQueryParams) => {
        const action = async () => await controller.getPayloads(queryParams);

        await expect(action()).rejects.toThrowError(BadRequestException);
      },
    );

    it('returns the expected valid response', async () => {
      const expectedResult = {
        pageNumber: 1,
        pageSize: 10,
        firstPage: '//payload?pageNumber=1&pageSize=10',
        lastPage: '//payload?pageNumber=1&pageSize=10',
        totalPages: 1,
        totalRecords: 3,
        nextPage: null,
        previousPage: null,
        data: [
          {
            payload_id: '86c0f117-4021-412e-b163-0dc621df672a',
            patient_id: '1d0253c4-8fab-41df-a414-55d52e4c6c3f',
            patient_name: 'Jane Doe',
            payload_received: '2022-08-17T12:21:10.203Z',
          },
          {
            payload_id: '30a8e0c6-e6c4-458f-aa4d-b224b493d3c0',
            patient_id: '',
            patient_name: '',
            payload_received: '2022-08-17T12:21:10.203Z',
          },
          {
            payload_id: 'c5c3636b-81dd-44a9-8c4b-71adec7d47b2',
            patient_id: 'fd1bebf4-d690-4fc5-a0d8-4fd4701ff4c9',
            patient_name: 'Steve Jobs',
            payload_received: '2022-08-17T12:21:10.2Z',
          },
        ],
      };

      service.getPayloads.mockResolvedValue(expectedResult);

      const response = await controller.getPayloads({
        pageNumber: '1',
        pageSize: '10',
      });

      expect(response).toStrictEqual(expectedResult);
    });
  });
});
