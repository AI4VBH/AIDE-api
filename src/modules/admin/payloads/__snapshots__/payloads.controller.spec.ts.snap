// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayloadsController getPayloadById returns the expected result 1`] = `
{
  "patient_id": "example_patient_id",
  "patient_name": "example name",
  "payload_id": "example_id",
  "payload_received": "received value",
}
`;
